### PR TITLE
Fix graph-mode evaluate() metric mismatch for fixed-batch tf.data datasets

### PR DESCRIPTION
**Pull request by maitry63**
This PR fixes incorrect metric values in `model.evaluate()` when running in graph mode (`run_eagerly=False`) on datasets created using` tf.data.Dataset.from_generato`r with a fixed batch size in the output_signature.

In these cases, stateful metrics (like MeanAbsoluteError) can return lower (diluted) values compared to eager mode.
This problem only happens in graph mode and does not appear when:
 - run_eagerly=True, or
 - using datasets with dynamic batch sizes (None), such as from_tensor_slices().batch(...)
 
The issue happens in the graph-mode evaluation path. During evaluation, TensorFlow uses a graph-based execution flow (tf.function) and loop logic to process the dataset.

For datasets with a fixed batch size, this graph execution can interact incorrectly with the iterator logic during tracing. As a result, metrics may get updated one extra time during the execution process.

This leads to:
 - higher metric counts than expected
 - lower (diluted) final metric values
 - mismatch between graph mode and eager mode results

Reference PR - [stinodego#1](https://github.com/stinodego/keras/pull/1)

Fixes: [#23084](https://github.com/keras-team/keras/issues/23084)

## Contributor Agreement

**Please review our [AI-Assisted Contribution Policy](../CONTRIBUTING.md#ai-assisted-contribution-policy) and check all boxes below before submitting your PR for review:**

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

**Note:** Failing to adhere to this agreement may result in your future PRs no longer being reviewed.

### CODE REVIEW

File: keras/src/backend/tensorflow/trainer.py [hunk 1]
```diff
@@ -154,7 +154,10 @@ def multi_step_on_iterator(iterator):
                     one_step_on_data(iterator.get_next())
                 )
```
Comment (hertschuh) [inline keras/src/backend/tensorflow/trainer.py]: Remove. This is now dead code.

File: keras/src/backend/tensorflow/trainer.py [hunk 2]
```diff
@@ -154,7 +154,10 @@ def multi_step_on_iterator(iterator):
...
 
-            # the spec is set lazily during the tracing of `tf.while_loop`
+            # Delay initialization of the Optional element_spec until the first
+            # successful iteration. This avoids additional tracing when using
+            # datasets with fully-defined batch dimensions while preserving the
+            # correct output specification for subsequent iterations.
```
Comment (hertschuh) [inline keras/src/backend/tensorflow/trainer.py]: I don't understand any of this comment, and it seems wrong. Just undo and restore the previous comment.

File: keras/src/backend/tensorflow/trainer.py [hunk 3]
```diff
@@ -236,15 +239,29 @@ def body(execution_step, optional_outputs, next_optional_inputs):
...
             if isinstance(
-                iterator, (tf.data.Iterator, tf.distribute.DistributedIterator)
+                iterator,
+                (tf.data.Iterator, tf.distribute.DistributedIterator),
             ):
+                # Fast path for the common case.
+                # Avoid routing single-step execution
+                # through the multi-step iterator helper,
+                # which is only needed when
+                # steps_per_execution > 1.
```
Comment (hertschuh) [inline keras/src/backend/tensorflow/trainer.py]: Remove this comment, it's pretty clear what's going on.

File: keras/src/trainers/trainer_test.py [hunk 4]
```diff
@@ -1765,6 +1765,68 @@ def test_evaluate_with_different_batch_size_same_loss(self):
...
+        self.assertAllClose(
+            graph_result["mae"]
+            if isinstance(graph_result, dict)
+            else graph_result,
+            eager_result["mae"]
+            if isinstance(eager_result, dict)
+            else eager_result,
+            atol=1e-5,
+            rtol=1e-5,
+        )
```
Comment (hertschuh) [inline keras/src/trainers/trainer_test.py:1812]: Rewrap this code. It could be way fewer lines of code if everything that can fit one line was put on one line.

File: keras/src/backend/tensorflow/trainer.py [hunk 5]
```diff
@@ -236,15 +239,29 @@ def body(execution_step, optional_outputs, next_optional_inputs):
 
         def function(iterator):
             if isinstance(
-                iterator, (tf.data.Iterator, tf.distribute.DistributedIterator)
+                iterator,
+                (tf.data.Iterator, tf.distribute.DistributedIterator),
```
Comment (hertschuh) [inline keras/src/backend/tensorflow/trainer.py]: Spurious re-wrapping, undo

**Review by hertschuh — CHANGES_REQUESTED** [review]
Thanks for looking into this!

File: keras/src/backend/tensorflow/trainer.py (hunk 1 above)
Comment (maitry63) [reply keras/src/backend/tensorflow/trainer.py]: Hi @hertschuh , thank you so much for your useful review. 
I fixed the changes.

File: keras/src/backend/tensorflow/trainer.py (hunk 2 above)
Comment (maitry63) [reply keras/src/backend/tensorflow/trainer.py]: Corrected

File: keras/src/backend/tensorflow/trainer.py (hunk 3 above)
Comment (maitry63) [reply keras/src/backend/tensorflow/trainer.py]: Corrected

File: keras/src/trainers/trainer_test.py (hunk 4 above)
Comment (maitry63) [reply keras/src/trainers/trainer_test.py:1812]: Updated

File: keras/src/backend/tensorflow/trainer.py (hunk 5 above)
Comment (maitry63) [reply keras/src/backend/tensorflow/trainer.py]: This rewrapping happened automatically after running 'black' for the formatting checks. Restored correctly.

Comment (hertschuh) [reply keras/src/backend/tensorflow/trainer.py]: Oh, we don't use `black` on the keras repo. We use `ruff`. This happens automatically on pre-commit.

https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md#generating-public-api-and-formatting-the-code

Comment (hertschuh) [reply keras/src/backend/tensorflow/trainer.py]: In fact the code format check is failing. Can you reformat?

File: keras/src/backend/tensorflow/trainer.py [hunk 6]
```diff
@@ -236,15 +231,22 @@ def body(execution_step, optional_outputs, next_optional_inputs):
 
         def function(iterator):
             if isinstance(
-                iterator, (tf.data.Iterator, tf.distribute.DistributedIterator)
+                iterator,(tf.data.Iterator, tf.distribute.DistributedIterator),
```
Comment (hertschuh) [inline keras/src/backend/tensorflow/trainer.py]: This formatting is incorrect: missing space and trailing comma.

**Review by hertschuh — CHANGES_REQUESTED** [review]

File: keras/src/backend/tensorflow/trainer.py (hunk 6 above)
Comment (maitry63) [reply keras/src/backend/tensorflow/trainer.py]: Corrected

File: keras/src/backend/tensorflow/trainer.py [hunk 7]
```diff
@@ -238,13 +233,20 @@ def function(iterator):
...
                 opt_outputs = multi_step_on_iterator(iterator)
                 if not opt_outputs.has_value():
                     raise StopIteration
                 return opt_outputs.get_value()
             else:
-                for step, data in zip(
-                    range(self.steps_per_execution), iterator
+                for _, data in zip(
+                    range(self.steps_per_execution),
+                    iterator,
```
Comment (stinodego) [inline keras/src/backend/tensorflow/trainer.py]: Apologies for inserting myself into the review process, but I noticed this unrelated formatting change is still in the diff.  Since hertschuh specifically requested to undo the spurious rewrapping it's probably best to undo this:

```suggestion
                for step, data in zip(
                    range(self.steps_per_execution), iterator
```

File: keras/src/trainers/trainer_test.py [hunk 8]
```diff
@@ -1765,6 +1765,52 @@ def test_evaluate_with_different_batch_size_same_loss(self):
...
+        eager_result = model_eager.evaluate(
+            make_dataset(), verbose=0, return_dict=True
+        )
+
+        self.assertAllClose(
+            graph_result["mae"],
+            eager_result["mae"],
+            atol=1e-5,
+            rtol=1e-5,
+        )
```
Comment (stinodego) [inline keras/src/trainers/trainer_test.py:1812]: Also according to hertschuh's remark, this statement should be on a single line:

```suggestion
        self.assertAllClose(
            graph_result["mae"], eager_result["mae"], atol=1e-5, rtol=1e-5
        )
```

Otherwise all seems good to go!

Comment (maitry63) [reply keras/src/trainers/trainer_test.py:1812]: Thank you for your review ! Corrected.

File: keras/src/backend/tensorflow/trainer.py (hunk 7 above)
Comment (maitry63) [reply keras/src/backend/tensorflow/trainer.py]: Thank you for your review ! Corrected.

**Review by hertschuh — APPROVED** [review]
Thanks for the fix!
